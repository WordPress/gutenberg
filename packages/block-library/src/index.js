/**
 * WordPress dependencies
 */
import {
	setDefaultBlockName,
	setFreeformContentHandlerName,
	setUnregisteredTypeHandlerName,
	setGroupingBlockName,
} from '@wordpress/blocks';

/**
 * Internal dependencies
 */
// When IS_GUTENBERG_PLUGIN is set to false, imports of experimental blocks
// are transformed by packages/block-library/src/index.js as follows:
//    import * as experimentalBlock from './experimental-block'
// becomes
//    const experimentalBlock = null;
// This enables webpack to eliminate the experimental blocks code from the
// production build to make the final bundle smaller.
//
// See https://github.com/WordPress/gutenberg/pull/40655 for more context.
import * as archives from './archives';
import * as avatar from './avatar';
import * as audio from './audio';
import * as button from './button';
import * as buttons from './buttons';
import * as calendar from './calendar';
import * as categories from './categories';
import * as classic from './freeform';
import * as code from './code';
import * as column from './column';
import * as columns from './columns';
import * as comments from './comments';
import * as commentAuthorAvatar from './comment-author-avatar';
import * as commentAuthorName from './comment-author-name';
import * as commentContent from './comment-content';
import * as commentDate from './comment-date';
import * as commentEditLink from './comment-edit-link';
import * as commentReplyLink from './comment-reply-link';
import * as commentTemplate from './comment-template';
import * as commentsPaginationPrevious from './comments-pagination-previous';
import * as commentsPagination from './comments-pagination';
import * as commentsPaginationNext from './comments-pagination-next';
import * as commentsPaginationNumbers from './comments-pagination-numbers';
import * as commentsTitle from './comments-title';
import * as cover from './cover';
import * as details from './details';
import * as embed from './embed';
import * as file from './file';
import * as form from './form';
import * as formInput from './form-input';
import * as formSubmitButton from './form-submit-button';
import * as formSubmissionNotification from './form-submission-notification';
import * as gallery from './gallery';
import * as group from './group';
import * as heading from './heading';
import * as homeLink from './home-link';
import * as html from './html';
import * as image from './image';
import * as latestComments from './latest-comments';
import * as latestPosts from './latest-posts';
import * as list from './list';
import * as listItem from './list-item';
import * as logInOut from './loginout';
import * as mediaText from './media-text';
import * as missing from './missing';
import * as more from './more';
import * as navigation from './navigation';
import * as navigationLink from './navigation-link';
import * as navigationSubmenu from './navigation-submenu';
import * as nextpage from './nextpage';
import * as pattern from './pattern';
import * as pageList from './page-list';
import * as pageListItem from './page-list-item';
import * as paragraph from './paragraph';
import * as postAuthor from './post-author';
import * as postAuthorName from './post-author-name';
import * as postAuthorBiography from './post-author-biography';
import * as postComment from './post-comment';
import * as postCommentsCount from './post-comments-count';
import * as postCommentsForm from './post-comments-form';
import * as postCommentsLink from './post-comments-link';
import * as postContent from './post-content';
import * as postDate from './post-date';
import * as postExcerpt from './post-excerpt';
import * as postFeaturedImage from './post-featured-image';
import * as postNavigationLink from './post-navigation-link';
import * as postTemplate from './post-template';
import * as postTerms from './post-terms';
import * as postTimeToRead from './post-time-to-read';
import * as postTitle from './post-title';
import * as preformatted from './preformatted';
import * as pullquote from './pullquote';
import * as query from './query';
import * as queryNoResults from './query-no-results';
import * as queryPagination from './query-pagination';
import * as queryPaginationNext from './query-pagination-next';
import * as queryPaginationNumbers from './query-pagination-numbers';
import * as queryPaginationPrevious from './query-pagination-previous';
import * as queryTitle from './query-title';
import * as quote from './quote';
import * as reusableBlock from './block';
import * as readMore from './read-more';
import * as rss from './rss';
import * as search from './search';
import * as separator from './separator';
import * as shortcode from './shortcode';
import * as siteLogo from './site-logo';
import * as siteTagline from './site-tagline';
import * as siteTitle from './site-title';
import * as socialLink from './social-link';
import * as socialLinks from './social-links';
import * as spacer from './spacer';
import * as table from './table';
import * as tableOfContents from './table-of-contents';
import * as tagCloud from './tag-cloud';
import * as templatePart from './template-part';
import * as termDescription from './term-description';
import * as textColumns from './text-columns';
import * as verse from './verse';
import * as video from './video';
import * as footnotes from './footnotes';

import isBlockMetadataExperimental from './utils/is-block-metadata-experimental';

/**
 * Function to get all the block-library blocks in an array
 */
const getAllBlocks = () => {
	const blocks = [
		// Common blocks are grouped at the top to prioritize their display
		// in various contexts â€” like the inserter and auto-complete components.
		paragraph,
		image,
		heading,
		gallery,
		list,
		listItem,
		quote,

		// Register all remaining core blocks.
		archives,
		audio,
		button,
		buttons,
		calendar,
		categories,
		code,
		column,
		columns,
		commentAuthorAvatar,
		cover,
		details,
		embed,
		file,
		group,
		html,
		latestComments,
		latestPosts,
		mediaText,
		missing,
		more,
		nextpage,
		pageList,
		pageListItem,
		pattern,
		preformatted,
		pullquote,
		reusableBlock,
		rss,
		search,
		separator,
		shortcode,
		socialLink,
		socialLinks,
		spacer,
		table,
		tagCloud,
		textColumns,
		verse,
		video,
		footnotes,

		// theme blocks
		navigation,
		navigationLink,
		navigationSubmenu,
		siteLogo,
		siteTitle,
		siteTagline,
		query,
		templatePart,
		avatar,
		postTitle,
		postExcerpt,
		postFeaturedImage,
		postContent,
		postAuthor,
		postAuthorName,
		postComment,
		postCommentsCount,
		postCommentsLink,
		postDate,
		postTerms,
		postNavigationLink,
		postTemplate,
		postTimeToRead,
		queryPagination,
		queryPaginationNext,
		queryPaginationNumbers,
		queryPaginationPrevious,
		queryNoResults,
		readMore,
		comments,
		commentAuthorName,
		commentContent,
		commentDate,
		commentEditLink,
		commentReplyLink,
		commentTemplate,
		commentsTitle,
		commentsPagination,
		commentsPaginationNext,
		commentsPaginationNumbers,
		commentsPaginationPrevious,
		postCommentsForm,
		tableOfContents,
		homeLink,
		logInOut,
		termDescription,
		queryTitle,
		postAuthorBiography,
	];
	if ( window?.__experimentalEnableFormBlocks ) {
		blocks.push( form );
		blocks.push( formInput );
		blocks.push( formSubmitButton );
		blocks.push( formSubmissionNotification );
	}

	// When in a WordPress context, conditionally
	// add the classic block and TinyMCE editor
	// under any of the following conditions:
	//   - the current post contains a classic block
	//   - the experiment to disable TinyMCE isn't active.
	//   - a query argument specifies that TinyMCE should be loaded
	if (
		window?.wp?.oldEditor &&
		( window?.wp?.needsClassicBlock ||
			! window?.__experimentalDisableTinymce ||
			!! new URLSearchParams( window?.location?.search ).get(
				'requiresTinymce'
			) )
	) {
		blocks.push( classic );
	}

	return blocks.filter( Boolean );
};

/**
 * Function to get all the core blocks in an array.
 *
 * @example
 * ```js
 * import { __experimentalGetCoreBlocks } from '@wordpress/block-library';
 *
 * const coreBlocks = __experimentalGetCoreBlocks();
 * ```
 */
export const __experimentalGetCoreBlocks = () =>
	getAllBlocks().filter(
		( { metadata } ) => ! isBlockMetadataExperimental( metadata )
	);

/**
 * Function to register core blocks provided by the block editor.
 *
 * @param {Array} blocks An optional array of the core blocks being registered.
 *
 * @example
 * ```js
 * import { registerCoreBlocks } from '@wordpress/block-library';
 *
 * registerCoreBlocks();
 * ```
 */
export const registerCoreBlocks = (
	blocks = __experimentalGetCoreBlocks()
) => {
	blocks.forEach( ( { init } ) => init() );

	setDefaultBlockName( paragraph.name );
	if (
		window.wp &&
		window.wp.oldEditor &&
		blocks.some( ( { name } ) => name === classic.name )
	) {
		setFreeformContentHandlerName( classic.name );
	}
	setUnregisteredTypeHandlerName( missing.name );
	setGroupingBlockName( group.name );
};

/**
 * Function to register experimental core blocks depending on editor settings.
 *
 * @param {boolean} enableFSEBlocks Whether to enable the full site editing blocks.
 * @example
 * ```js
 * import { __experimentalRegisterExperimentalCoreBlocks } from '@wordpress/block-library';
 *
 * __experimentalRegisterExperimentalCoreBlocks( settings );
 * ```
 */
export const __experimentalRegisterExperimentalCoreBlocks =
	globalThis.IS_GUTENBERG_PLUGIN
		? ( { enableFSEBlocks } = {} ) => {
				const enabledExperiments = [ enableFSEBlocks ? 'fse' : null ];
				getAllBlocks()
					.filter( ( { metadata } ) =>
						isBlockMetadataExperimental( metadata )
					)
					.filter(
						( { metadata: { __experimental } } ) =>
							__experimental === true ||
							enabledExperiments.includes( __experimental )
					)
					.forEach( ( { init } ) => init() );
		  }
		: undefined;

export { privateApis } from './private-apis';
