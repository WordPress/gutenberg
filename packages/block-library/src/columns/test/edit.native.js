/**
 * External dependencies
 */
import {
	act,
	addBlock,
	dismissModal,
	fireEvent,
	getBlock,
	getEditorHtml,
	initializeEditor,
	openBlockActionsMenu,
	openBlockSettings,
	screen,
	waitForModalVisible,
	within,
} from 'test/helpers';

/**
 * WordPress dependencies
 */
import { getBlockTypes, unregisterBlockType } from '@wordpress/blocks';
import { registerCoreBlocks } from '@wordpress/block-library';

const TWO_COLUMNS_BLOCK_HTML = `<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->
<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->`;

beforeAll( () => {
	// Register all core blocks
	registerCoreBlocks();
} );

afterAll( () => {
	// Clean up registered blocks
	getBlockTypes().forEach( ( block ) => {
		unregisterBlockType( block.name );
	} );
} );

describe( 'Columns block', () => {
	it( 'inserts block', async () => {
		await initializeEditor();

		// Add block
		await addBlock( screen, 'Columns' );

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		expect( columnsBlock ).toBeVisible();
		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	it( 'adds a column block using the appender', async () => {
		await initializeEditor( {
			initialHtml: TWO_COLUMNS_BLOCK_HTML,
		} );

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Add a new column
		const appenderButton =
			within( columnsBlock ).getByTestId( 'appender-button' );
		fireEvent.press( appenderButton );

		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	describe( 'when using the number of columns setting', () => {
		it( 'adds a column block when incrementing the value', async () => {
			await initializeEditor( {
				initialHtml: TWO_COLUMNS_BLOCK_HTML,
			} );
			const { getByLabelText } = screen;

			// Get block
			const columnsBlock = await getBlock( screen, 'Columns' );
			fireEvent.press( columnsBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Update the number of columns by adding one
			const columnsControl = getByLabelText( /Number of columns/ );
			fireEvent( columnsControl, 'accessibilityAction', {
				nativeEvent: { actionName: 'increment' },
			} );

			expect( getEditorHtml() ).toMatchSnapshot();
		} );

		it( 'adds at least 15 Column blocks without limitation', async () => {
			await initializeEditor( {
				initialHtml: TWO_COLUMNS_BLOCK_HTML,
			} );
			const { getByLabelText } = screen;

			// Get block
			const columnsBlock = await getBlock( screen, 'Columns' );
			fireEvent.press( columnsBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Update the number of columns
			const columnsControl = getByLabelText( /Number of columns/ );

			for ( let x = 0; x < 15; x++ ) {
				fireEvent( columnsControl, 'accessibilityAction', {
					nativeEvent: { actionName: 'increment' },
				} );
			}

			expect( getEditorHtml() ).toMatchSnapshot();
		} );

		it( 'removes a column block when decrementing the value', async () => {
			await initializeEditor( {
				initialHtml: TWO_COLUMNS_BLOCK_HTML,
			} );
			const { getByLabelText } = screen;

			// Wait for the block to be created.
			const columnsBlock = await getBlock( screen, 'Columns' );
			fireEvent.press( columnsBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Update the number of columns by removing one
			const columnsControl = getByLabelText( /Number of columns/ );
			fireEvent( columnsControl, 'accessibilityAction', {
				nativeEvent: { actionName: 'decrement' },
			} );

			expect( getEditorHtml() ).toMatchSnapshot();
		} );

		it( 'reaches the minimum limit of number of column blocks', async () => {
			await initializeEditor();

			// Add block
			await addBlock( screen, 'Columns' );

			// Wait for the variations modal to be visible
			const blockVariationModal = await screen.findByTestId(
				'block-variation-modal'
			);
			await waitForModalVisible( blockVariationModal );

			// Select a column variation
			const threeColumnLayout =
				within( blockVariationModal ).getByLabelText(
					/33 \/ 33 \/ 33 block/
				);
			fireEvent.press( threeColumnLayout );

			// Get block
			const columnsBlock = await getBlock( screen, 'Columns' );
			fireEvent.press( columnsBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Update the number of columns by adding one
			const columnsControl = screen.getByLabelText( /Number of columns/ );
			fireEvent( columnsControl, 'accessibilityAction', {
				nativeEvent: { actionName: 'increment' },
			} );

			// Press the decrement button 5 times to remove all columns but one
			for ( let index = 0; index < 5; index++ ) {
				fireEvent( columnsControl, 'accessibilityAction', {
					nativeEvent: { actionName: 'decrement' },
				} );
			}

			expect( getEditorHtml() ).toMatchSnapshot();
		} );
	} );

	it( 'removes column with the remove button', async () => {
		await initializeEditor( {
			initialHtml: TWO_COLUMNS_BLOCK_HTML,
		} );
		const { getByLabelText } = screen;

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Get the first column
		const firstColumnBlock = await getBlock( screen, 'Column' );
		fireEvent.press( firstColumnBlock );

		// Open block actions menu
		const blockActionsButton = getByLabelText( /Open Block Actions Menu/ );
		fireEvent.press( blockActionsButton );

		// Delete block
		const deleteButton = getByLabelText( /Remove block/ );
		fireEvent.press( deleteButton );

		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	it( 'removes the only one left Column with the remove button', async () => {
		await initializeEditor( {
			initialHtml: TWO_COLUMNS_BLOCK_HTML,
		} );
		const { getByLabelText } = screen;

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Get the first column
		const firstColumnBlock = await getBlock( screen, 'Column' );
		fireEvent.press( firstColumnBlock );

		// Open block actions menu
		let blockActionsButton = getByLabelText( /Open Block Actions Menu/ );
		fireEvent.press( blockActionsButton );

		// Delete block
		let deleteButton = getByLabelText( /Remove block/ );
		fireEvent.press( deleteButton );

		// Get the only left column
		const lastColumnBlock = await getBlock( screen, 'Column' );
		fireEvent.press( lastColumnBlock );

		// Open block actions menu
		blockActionsButton = getByLabelText( /Open Block Actions Menu/ );
		fireEvent.press( blockActionsButton );

		// Delete block
		deleteButton = getByLabelText( /Remove block/ );
		fireEvent.press( deleteButton );

		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	it( 'changes vertical alignment on Columns', async () => {
		await initializeEditor( {
			initialHtml: TWO_COLUMNS_BLOCK_HTML,
		} );
		const { getByLabelText } = screen;

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Open vertical alignment menu
		const verticalAlignmentButton = getByLabelText(
			/Change vertical alignment/
		);
		fireEvent.press( verticalAlignmentButton );

		// Get Align top button
		const verticalTopAlignmentButton = getByLabelText( /Align top/ );
		fireEvent.press( verticalTopAlignmentButton );

		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	it( 'changes the vertical alignment on individual Column', async () => {
		await initializeEditor( {
			initialHtml: TWO_COLUMNS_BLOCK_HTML,
		} );

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Open vertical alignment menu
		const verticalAlignmentButton = screen.getByLabelText(
			/Change vertical alignment/
		);
		fireEvent.press( verticalAlignmentButton );

		// Get Align top button
		const verticalTopAlignmentButton = screen.getByLabelText( /Align top/ );
		fireEvent.press( verticalTopAlignmentButton );

		// Get the first column
		const firstColumnBlock = await getBlock( screen, 'Column' );
		fireEvent.press( firstColumnBlock );
		// Inner blocks batch store updates with microtasks.
		// To avoid `act` warnings, we let queued microtasks to be executed.
		// Reference: https://t.ly/b95nA
		await act( async () => {} );

		// Open vertical alignment menu
		fireEvent.press( verticalAlignmentButton );

		// Get Align bottom button
		const verticalBottomAlignmentButton =
			screen.getByLabelText( /Align bottom/ );
		fireEvent.press( verticalBottomAlignmentButton );

		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	it( 'sets current vertical alignment on new Columns', async () => {
		await initializeEditor( {
			initialHtml: TWO_COLUMNS_BLOCK_HTML,
		} );
		const { getByLabelText } = screen;

		// Get block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Open vertical alignment menu
		const verticalAlignmentButton = getByLabelText(
			/Change vertical alignment/
		);
		fireEvent.press( verticalAlignmentButton );

		// Get Align top button
		const verticalTopAlignmentButton = getByLabelText( /Align top/ );
		fireEvent.press( verticalTopAlignmentButton );

		// Add a new column
		const appenderButton =
			within( columnsBlock ).getByTestId( 'appender-button' );
		fireEvent.press( appenderButton );

		expect( getEditorHtml() ).toMatchSnapshot();
	} );

	describe( 'when using columns percentage mechanism', () => {
		it( "updates the slider's input value", async () => {
			await initializeEditor();

			// Add block
			await addBlock( screen, 'Columns' );

			// Wait for the variations modal to be visible
			const blockVariationModal = await screen.findByTestId(
				'block-variation-modal'
			);
			await waitForModalVisible( blockVariationModal );

			// Select a column variation
			const threeColumnLayout =
				within( blockVariationModal ).getByLabelText(
					/33 \/ 33 \/ 33 block/
				);
			fireEvent.press( threeColumnLayout );

			// Get the first column
			const firstColumnBlock = await getBlock( screen, 'Column' );
			fireEvent.press( firstColumnBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Get width control
			const widthControl = screen.getByLabelText( /Width. Value is/ );
			fireEvent.press(
				within( widthControl ).getByText( '33.3', { hidden: true } )
			);
			const widthTextInput = within( widthControl ).getByDisplayValue(
				'33.3',
				{ hidden: true }
			);
			fireEvent.changeText( widthTextInput, '55.55555' );

			expect( getEditorHtml() ).toMatchSnapshot();
		} );

		it( 'sets custom values correctly', async () => {
			await initializeEditor( {
				initialHtml: TWO_COLUMNS_BLOCK_HTML,
			} );
			const { getByLabelText, getByTestId } = screen;

			// Get block
			const columnsBlock = await getBlock( screen, 'Columns' );
			fireEvent.press( columnsBlock );

			// Get the first column
			const firstColumnBlock = await getBlock( screen, 'Column' );
			fireEvent.press( firstColumnBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Set custom width value for the first column
			let widthControl = getByLabelText( /Width. Value is/ );
			fireEvent.press(
				within( widthControl ).getByText( '50', { hidden: true } )
			);
			let widthTextInput = within( widthControl ).getByDisplayValue(
				'50',
				{ hidden: true }
			);
			fireEvent.changeText( widthTextInput, '90' );

			// Dismiss settings
			await dismissModal( getByTestId( 'block-settings-modal' ) );

			// Get the Second column
			const secondColumnBlock = await getBlock( screen, 'Column', {
				rowIndex: 2,
			} );
			fireEvent.press( secondColumnBlock );

			// Open block settings
			await openBlockSettings( screen );

			// Set custom width value for the second column
			widthControl = getByLabelText( /Width. Value is/ );
			fireEvent.press(
				within( widthControl ).getByText( '50', { hidden: true } )
			);
			widthTextInput = within( widthControl ).getByDisplayValue( '50', {
				hidden: true,
			} );
			fireEvent.changeText( widthTextInput, '55.5' );

			expect( getEditorHtml() ).toMatchSnapshot();
		} );
	} );

	describe( 'when using the columns layout picker', () => {
		const testData = [
			[ '100 block' ],
			[ '50 / 50 block' ],
			[ '33 / 66 block' ],
			[ '66 / 33 block' ],
			[ '33 / 33 / 33 block' ],
			[ '25 / 50 / 25 block' ],
		];

		test.each( testData )(
			'sets the predefined percentages for %s',
			async ( layout ) => {
				await initializeEditor();

				// Add block
				await addBlock( screen, 'Columns' );

				// Wait for the variations modal to be visible
				const blockVariationModal = await screen.findByTestId(
					'block-variation-modal'
				);
				await waitForModalVisible( blockVariationModal );

				// Select a column variation
				const columnLayout =
					within( blockVariationModal ).getByLabelText( layout );
				fireEvent.press( columnLayout );

				expect( getEditorHtml() ).toMatchSnapshot();
			}
		);
	} );

	it( 'transforms a nested Columns block into a Group block', async () => {
		await initializeEditor( {
			initialHtml: `<!-- wp:group {"layout":{"type":"constrained"}} -->
			<div class="wp-block-group"><!-- wp:columns -->
			<div class="wp-block-columns"><!-- wp:column {"width":"100%"} -->
			<div class="wp-block-column" style="flex-basis:100%"><!-- wp:paragraph -->
			<p></p>
			<!-- /wp:paragraph --></div>
			<!-- /wp:column --></div>
			<!-- /wp:columns --></div>
			<!-- /wp:group -->`,
		} );

		// Get Columns block
		const columnsBlock = await getBlock( screen, 'Columns' );
		fireEvent.press( columnsBlock );

		// Open block actions menu
		await openBlockActionsMenu( screen );

		// Tap on the Transform block button
		fireEvent.press( screen.getByLabelText( /Transform block…/ ) );

		// Tap on the Group transform button
		fireEvent.press( screen.getByLabelText( 'Group' ) );
		expect( getEditorHtml() ).toMatchSnapshot();
	} );
} );
