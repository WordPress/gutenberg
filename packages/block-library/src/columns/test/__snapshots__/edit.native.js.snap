// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns block adds a column block using the appender 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block changes the vertical alignment on individual Column 1`] = `
"<!-- wp:columns {"verticalAlignment":"top"} -->
<div class="wp-block-columns are-vertically-aligned-top"><!-- wp:column {"verticalAlignment":"bottom"} -->
<div class="wp-block-column is-vertically-aligned-bottom"></div>
<!-- /wp:column -->

<!-- wp:column {"verticalAlignment":"top"} -->
<div class="wp-block-column is-vertically-aligned-top"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block changes vertical alignment on Columns 1`] = `
"<!-- wp:columns {"verticalAlignment":"top"} -->
<div class="wp-block-columns are-vertically-aligned-top"><!-- wp:column {"verticalAlignment":"top"} -->
<div class="wp-block-column is-vertically-aligned-top"></div>
<!-- /wp:column -->

<!-- wp:column {"verticalAlignment":"top"} -->
<div class="wp-block-column is-vertically-aligned-top"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block inserts block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"></div>
<!-- /wp:column -->

<!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block removes column with the remove button 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block removes the only one left Column with the remove button 1`] = `
"<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->"
`;

exports[`Columns block sets current vertical alignment on new Columns 1`] = `
"<!-- wp:columns {"verticalAlignment":"top"} -->
<div class="wp-block-columns are-vertically-aligned-top"><!-- wp:column {"verticalAlignment":"top"} -->
<div class="wp-block-column is-vertically-aligned-top"></div>
<!-- /wp:column -->

<!-- wp:column {"verticalAlignment":"top"} -->
<div class="wp-block-column is-vertically-aligned-top"></div>
<!-- /wp:column -->

<!-- wp:column {"verticalAlignment":"top"} -->
<div class="wp-block-column is-vertically-aligned-top"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block transforms a nested Columns block into a Group block 1`] = `
"<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"100%"} -->
<div class="wp-block-column" style="flex-basis:100%"><!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->"
`;

exports[`Columns block when using columns percentage mechanism sets custom values correctly 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"90%"} -->
<div class="wp-block-column" style="flex-basis:90%"></div>
<!-- /wp:column -->

<!-- wp:column {"width":"55.5%"} -->
<div class="wp-block-column" style="flex-basis:55.5%"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using columns percentage mechanism updates the slider's input value 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"55.6%"} -->
<div class="wp-block-column" style="flex-basis:55.6%"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the columns layout picker sets the predefined percentages for 25 / 50 / 25 block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"25%"} -->
<div class="wp-block-column" style="flex-basis:25%"></div>
<!-- /wp:column -->

<!-- wp:column {"width":"50%"} -->
<div class="wp-block-column" style="flex-basis:50%"></div>
<!-- /wp:column -->

<!-- wp:column {"width":"25%"} -->
<div class="wp-block-column" style="flex-basis:25%"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the columns layout picker sets the predefined percentages for 33 / 33 / 33 block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the columns layout picker sets the predefined percentages for 33 / 66 block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"33.33%"} -->
<div class="wp-block-column" style="flex-basis:33.33%"></div>
<!-- /wp:column -->

<!-- wp:column {"width":"66.66%"} -->
<div class="wp-block-column" style="flex-basis:66.66%"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the columns layout picker sets the predefined percentages for 50 / 50 block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the columns layout picker sets the predefined percentages for 66 / 33 block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"width":"66.66%"} -->
<div class="wp-block-column" style="flex-basis:66.66%"></div>
<!-- /wp:column -->

<!-- wp:column {"width":"33.33%"} -->
<div class="wp-block-column" style="flex-basis:33.33%"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the columns layout picker sets the predefined percentages for 100 block 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the number of columns setting adds a column block when incrementing the value 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the number of columns setting adds at least 15 Column blocks without limitation 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the number of columns setting reaches the minimum limit of number of column blocks 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;

exports[`Columns block when using the number of columns setting removes a column block when decrementing the value 1`] = `
"<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->"
`;
